### PR TITLE
rust: remove call to `bindings::errname` when building `testlib`

### PR DESCRIPTION
**Pull request by wedsonaf**
This allows us to use, for example, `Result::unwrap` from example code
and still be able to compile, link, and run the example.

Without this change, an attempt to use `unwrap` in an example results in
the following error when making `rusttest`:

```
rust/kernel/error.rs:360: undefined reference to `errname'

= help: some `extern` functions couldn't be found; some native libraries
may need to be installed or have their path specified
```

Signed-off-by: Wedson Almeida Filho <wedsonaf@google.com>

### CODE REVIEW

**Comment by ojeda** [comment]
Sounds good to me.

**Comment by bjorn3** [comment]
Shouldn't this use `#[cfg]` instead of `cfg!()` Otherwise `bindings::errname` may still be referenced in debug builds. (MIR optimizations may eliminate it even in debug mode, though for example miri disables MIR optimizations to catch more UB)

**Comment by wedsonaf** [comment]
> Shouldn't this use `#[cfg]` instead of `cfg!()` Otherwise `bindings::errname` may still be referenced in debug builds. (MIR optimizations may eliminate it even in debug mode, though for example miri disables MIR optimizations to catch more UB)

Probably better. I just changed it to use a function, as assigning attributes to statements didn't seem very readable, and assigning them to expressions is unstable.

PTAL.

File: rust/kernel/error.rs [hunk 1]
```diff
@@ -352,13 +352,22 @@ impl Error {
     pub fn to_kernel_errno(self) -> c_types::c_int {
         self.0
     }
+
+    #[cfg(not(testlib))]
+    fn errname(&self) -> *const c_types::c_char {
+        // SAFETY: Just an FFI call, there are no extra safety requirements.
+        unsafe { bindings::errname(-self.0) }
```
Comment (bjorn3) [inline rust/kernel/error.rs]: Does `errname` allow passing in invalid error codes?

Comment (wedsonaf) [reply rust/kernel/error.rs]: Yes, it does, it returns `NULL` in such cases.

You can see the code [here](https://elixir.bootlin.com/linux/latest/source/lib/errname.c#L217).

Comment (nbdd0121) [reply rust/kernel/error.rs]: We might just want to make the signature `fn errname(self) -> Option<&'static CStr>` or  `fn errname(self) -> Option<&'static str>`.

Comment (wedsonaf) [reply rust/kernel/error.rs]: > We might just want to make the signature `fn errname(self) -> Option<&'static CStr>` or `fn errname(self) -> Option<&'static str>`.

No, I added this indirection solely for the attribute: I want these functions to do as little as possible. Also, to make this return an `Option`, I'll have to check for `null` here, then check if the `Option` is `Some` or `None` in the caller.

Comment (nbdd0121) [reply rust/kernel/error.rs]: I mean it might be useful to expose `errname` to users anyway. Converting a pointer to `Option` isn't bad either.

Comment (wedsonaf) [reply rust/kernel/error.rs]: Oh, I see.

Yes, I think it would be helpful. I'll make this change.

**Comment by wedsonaf** [comment]
Made `name` public. PTAL.

File: rust/kernel/error.rs [hunk 2]
```diff
@@ -352,22 +352,42 @@ impl Error {
...
+    }
 }
 
 impl fmt::Debug for Error {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        // SAFETY: FFI call.
-        let name = unsafe { bindings::errname(-self.0) };
-
-        if name.is_null() {
+        let name = if let Some(n) = self.name() {
```
Comment (nbdd0121) [inline rust/kernel/error.rs]: Perhaps
```rust
match self.name() {
    Some(name) => {
        // SAFETY: These strings are ASCII-only.
        let str = unsafe { str::from_utf8_unchecked(name) };
        f.debug_tuple(str).finish()
    }
    None => {
        // Print out number if no name can be found.
        f.debug_tuple("Error").field(&-self.0).finish()
    }
}
```

Comment (wedsonaf) [reply rust/kernel/error.rs]: Sure, I changed to code to use a `match`.

**Review by bjorn3 — APPROVED** [review]

**Pull request merged**